Introduce properties for functions

rschu1ze opened the pull request:
This PR is preparation for a re-implementation of #55035 after it was removed with #66630.

In the previous implementation, system table `system.functions` relied on virtual function `IFunction::isDeterministic` to fill field `is_deterministic` in the system table. In order to call `IFunction::isDeterministic`, the function had to be instantiated. In certain situations, function instantiation could throw an exception, e.g. because a dictionary is not initialized. This is actually all good and expected, and the system table had a try-catch that would fill in a dummy value in this case. But then there was also stateless test `02815_no_throw_in_simple_queries` which runs a simple query under environment variable `CLICKHOUSE_TERMINATE_ON_ANY_EXCEPTION=1`. The test uses clickHouse-local which internally queries `system.functions` to provide SQL syntax highlighting and completion. This led to an exception thrown and caught internally, but the test still failed - I guess rightfully as simple queries should throw no exceptions at all (even if they are handled). #66630 then removed the previous implementation.

A future PR will re-implement `system.functions.is_deterministic` using *function properties*, similar to what we already have for table functions (--> table function properties) and aggregation functions (--> aggregation function properties). These can be queried without instantiating the function.

This PR introduces the necessary infrastructure for this but does not add actually function properties yet.

Reviewer note: the individual commits are small, self-contained, and they have commit messages. For mental sanity, better check them instead of checking the whole PR at once.

### Changelog category (leave one):
- Not for changelog (changelog entry is not required)

robot-clickhouse commented:
<!-- automatic status comment for PR #66697 from rschu1ze/ClickHouse:function-factory-properties -->
*This is an automated comment for commit c2b00e0e23c9229a29f1bafb092d253a76d24bc2 with description of existing statuses. It's updated for the latest CI running*

[✅ Click here](https://s3.amazonaws.com/clickhouse-test-reports/66697/c2b00e0e23c9229a29f1bafb092d253a76d24bc2/ci_running.html) to open a full report in a separate page

<details><summary>Successful checks</summary>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>Style check</td><td>Runs a set of checks to keep the code style clean. If some of tests failed, see the related log from the report</td><td><a href="https://s3.amazonaws.com/clickhouse-test-reports/66697/c2b00e0e23c9229a29f1bafb092d253a76d24bc2/style_check.html">✅ success</a></td></tr>
<tbody>
</table>
</details>

rschu1ze committed:
TableFunctionFactory: Unnest documentation from TableFunctionProperties

Before this commit, FunctionDocumentation was a member of
TableFunctionProperties. This commit unnests it and moves it to the same
level as TableFunctionProperties.

Reasons:

- Documentation is no "property" which can be used to make runtime
  decisions. Its can only be output by some system tables.

- Function 'registerFunction' now puts more emphasis on documenting
  table functions, previously the syntax the documentation appeared
  as optional.

rschu1ze committed:
Add property infrastructure to FunctionFactory

Adds FunctionProperties to FunctionFactoryData, analogous to
TableFunctionProperties for TableFunctionFactoryData. No properties
are currently defined.

rschu1ze committed:
Cosmetics: Rename AggregateFunctionWithProperties to AggregateFunctionFactoryData

Makes the naming consistent with an equivalent structure in
TableFunctionFactory (TableFunctionFactoryData).

rschu1ze committed:
Cosmetics: Move function declarations out of header

rschu1ze committed:
Cosmetics: Parameter "doc" --> "documentation"

rschu1ze committed:
Cosmetics: Whitespace changes to group related functions together

rschu1ze committed:
Cosmetics: Move include to right header

rschu1ze committed:
Cosmetics: Sort includes

rschu1ze committed:
Cosmetics: Move functions around to keep things grouped

rschu1ze committed:
Cosmetics: Replace pair by struct

- more idiomatic access: .first/second vs. .<member>
- makes the code consistent with TableFunctionFactory

rschu1ze committed:
Cosmetics: Simplify FunctionFactory::tryGet

rschu1ze committed:
Cosmetics: Move function definition around

rschu1ze committed:
Cosmetics: Fix indentation

rschu1ze committed:
Cosmetics: Better variable naming

rschu1ze committed:
Cosmetics: Keeping simple things simple

rschu1ze committed:
Cosmetics: Trim parantheses and whitespaces

rschu1ze committed:
Cleanup interaction between IFactoryWithAliases and Factories

rschu1ze commented:
All tests destroyed.

rschu1ze commented:
One fine day.
